trio/_path.py

```diff
@@ -1,49 +1,86 @@
+from __future__ import annotations
+
+import inspect
 import os
 import pathlib
 import sys
 import types
+from collections.abc import Awaitable, Callable, Iterable
 from functools import partial, wraps
-from typing import TYPE_CHECKING, Any
+from io import BufferedRandom, BufferedReader, BufferedWriter, FileIO, TextIOWrapper
+from typing import IO, TYPE_CHECKING, Any, BinaryIO, ClassVar, TypeVar, cast, overload
 
 import trio
+from trio._file_io import AsyncIOWrapper as _AsyncIOWrapper
 from trio._util import Final, async_wraps
 
+if TYPE_CHECKING:
+    from _typeshed import (
+        OpenBinaryMode,
+        OpenBinaryModeReading,
+        OpenBinaryModeUpdating,
+        OpenBinaryModeWriting,
+        OpenTextMode,
+    )
+    from typing_extensions import Concatenate, Literal, ParamSpec, TypeAlias
+
+    P = ParamSpec("P")
+
+T = TypeVar("T")
+StrPath: TypeAlias = "str | os.PathLike[str]"
```

Comment:
Could you make this a `typing.Union` and not have to keep it as a string? I think that's the approach we've taken elsewhere. It would be nice to be consistent!